README.md
R implementation of the PhenoGraph algorithm
===============

### Rphenograph
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/TomKellyGenetics/Rphenograph?branch=master&svg=true)](https://ci.appveyor.com/project/TomKellyGenetics/Rphenograph)
[![Travis-CI Build Status](https://travis-ci.org/TomKellyGenetics/Rphenograph.svg?branch=master)](https://travis-ci.org/TomKellyGenetics/Rphenograph)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Coverage Status](https://img.shields.io/codecov/c/github/JinmiaoChenLab/Rphenograph/master.svg)](https://codecov.io/github/JinmiaoChenLab/Rphenograph?branch=master)

A simple R implementation of the [PhenoGraph](http://www.cell.com/cell/abstract/S0092-8674(15)00637-6) [1] algorithm, which is a clustering method designed for high-dimensional single-cell data analysis. It works by creating a graph ("network") representing phenotypic similarities between cells by calclating the Jaccard coefficient between nearest-neighbor sets, and then identifying communities using the well known [Louvain method](https://sites.google.com/site/findcommunities/) in this graph. 


### Installation

To install the latest version from the github repository, use:

``` r
if(!require(devtools)){
  install.packages("devtools") # If not already installed
}
devtools::install_github("i-cyto/Rphenograph")
```


### Usage

After installing the package, use the following code to run a simple example (to install, see below).

``` r
iris_unique <- unique(iris) # Remove duplicates
data <- as.matrix(iris_unique[,1:4])
Rphenograph_out <- Rphenograph(data, k = 45)
modularity(Rphenograph_out[[2]])
membership(Rphenograph_out[[2]])
iris_unique$phenograph_cluster <- factor(membership(Rphenograph_out[[2]]))
ggplot(iris_unique, aes(x=Sepal.Length, y=Sepal.Width, col=Species, shape=phenograph_cluster)) + geom_point(size = 3)+theme_bw()

```

![](Rhpenograph_iris_cluster.png)

### Reference

[1] Levine JH, Simonds EF, Bendall SC, Davis KL, Amir ED, Tadmor MD, et al. Data-Driven Phenotypic Dissection of AML Reveals Progenitor-like Cells that Correlate with Prognosis. Cell. Elsevier Inc.; 2015; 1–14. doi:10.1016/j.cell.2015.05.047




NEWS.md
# Rphenograph 0.99.1.9003


## New features

* S. Thomas Kelly added pruning and graphs clustering methods. This ahs been turned into calling igraph functions.

* Etienne K. Becht added approximate HNSW nearest neighbors for speed.

* S. Granjeaud improves the the Jaccard_coefficient function by pre-sorting nearest indices. Now the computation takes only a few seconds for a dataset of 300 k datapoints and 30 nearest neighbors.

* Louvain is the default graph clustering method. Any functions of the (r)igraph package can be specified.


## Bug fixes and minor improvements

* EK Becht reported that some points are missing in the grap. Self loop are added in the Jaccard C code.


# Rphenograph 0.99.1

Release from H. Chen who created and published this package.


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

// jaccard_coeff
NumericMatrix jaccard_coeff(NumericMatrix idx);
RcppExport SEXP _Rphenograph_jaccard_coeff(SEXP idxSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericMatrix >::type idx(idxSEXP);
    rcpp_result_gen = Rcpp::wrap(jaccard_coeff(idx));
    return rcpp_result_gen;
END_RCPP
}
// jaccard_coeff_1
NumericMatrix jaccard_coeff_1(NumericMatrix idx);
RcppExport SEXP _Rphenograph_jaccard_coeff_1(SEXP idxSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericMatrix >::type idx(idxSEXP);
    rcpp_result_gen = Rcpp::wrap(jaccard_coeff_1(idx));
    return rcpp_result_gen;
END_RCPP
}
// jaccard_coeff_3
NumericMatrix jaccard_coeff_3(NumericMatrix idx);
RcppExport SEXP _Rphenograph_jaccard_coeff_3(SEXP idxSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericMatrix >::type idx(idxSEXP);
    rcpp_result_gen = Rcpp::wrap(jaccard_coeff_3(idx));
    return rcpp_result_gen;
END_RCPP
}
// jaccard_coeff_4
NumericMatrix jaccard_coeff_4(NumericMatrix idx, IntegerVector ks);
RcppExport SEXP _Rphenograph_jaccard_coeff_4(SEXP idxSEXP, SEXP ksSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericMatrix >::type idx(idxSEXP);
    Rcpp::traits::input_parameter< IntegerVector >::type ks(ksSEXP);
    rcpp_result_gen = Rcpp::wrap(jaccard_coeff_4(idx, ks));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_Rphenograph_jaccard_coeff", (DL_FUNC) &_Rphenograph_jaccard_coeff, 1},
    {"_Rphenograph_jaccard_coeff_1", (DL_FUNC) &_Rphenograph_jaccard_coeff_1, 1},
    {"_Rphenograph_jaccard_coeff_3", (DL_FUNC) &_Rphenograph_jaccard_coeff_3, 1},
    {"_Rphenograph_jaccard_coeff_4", (DL_FUNC) &_Rphenograph_jaccard_coeff_4, 2},
    {NULL, NULL, 0}
};

RcppExport void R_init_Rphenograph(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


src/jaccard_coeff.cpp
#include <Rcpp.h>
using namespace Rcpp;

// Compute jaccard coefficient between nearest-neighbor sets
//
// Weights of both i->j and j->i are recorded if they have intersection. In this case
// w(i->j) should be equal to w(j->i). In some case i->j has weights while j<-i has no
// intersections, only w(i->j) is recorded. This is determinded in code `if(u>0)`. 
// In this way, the undirected graph is symmetrized by halfing the weight 
// in code `weights(r, 2) = u/(2.0*ncol - u)/2`.
//
// Author: Chen Hao, Date: 25/09/2015


// [[Rcpp::export]]
NumericMatrix jaccard_coeff(NumericMatrix idx) {
    int nrow = idx.nrow(), ncol = idx.ncol();
    NumericMatrix weights(nrow*ncol, 3);
    int r = 0;
    for (int i = 0; i < nrow; i++) {
        for (int j = 0; j < ncol; j++) {
            int k = idx(i,j)-1;
            NumericVector nodei = idx(i,_);
            NumericVector nodej = idx(k,_);
            int u = intersect(nodei, nodej).size();  // count intersection number
            if(u>0){ 
                weights(r, 0) = i+1;
                weights(r, 1) = k+1;
                weights(r, 2) = u/(2.0*ncol - u)/2;  // symmetrize the graph
                r++;
            }
        }
    }
    
    return weights;
}


// Slightly improve ressources usage
// Author: S. Granjeaud, Date: 17/02/2020

// [[Rcpp::export]]
NumericMatrix jaccard_coeff_1(NumericMatrix idx) {
    int nrow = idx.nrow(), ncol = idx.ncol();
    NumericMatrix weights(nrow*ncol, 3);
    int r = 0;
    NumericVector nodei (ncol);
    NumericVector nodej (ncol);
    for (int i = 0; i < nrow; i++) {
        nodei = idx(i,_);
        for (int j = 0; j < ncol; j++) {
            int k = idx(i,j)-1;
            nodej = idx(k,_);
            int u = intersect(nodei, nodej).size();  // count intersection number
            if (u > 0) {
                weights(r, 0) = i+1;
                weights(r, 1) = k+1;
                weights(r, 2) = u/(4.0*ncol - 2*u);  // symmetrize the graph
                r++;
            }
        }
    }
    
    return weights;
}


// Improve intersection computation by pre-sorting the indices, what requires more memory
// Author: S. Granjeaud, Date: 17/02/2020

// [[Rcpp::export]]
NumericMatrix jaccard_coeff_3(NumericMatrix idx) {
    int nrow = idx.nrow(), ncol = idx.ncol();
    NumericMatrix weights(nrow*ncol, 3);
    int r = 0;
    // allocate memory
    int** i_idx = new int*[nrow];
    if (nrow) {
        i_idx[0] = new int[nrow * ncol];  // mem alloc
        for (int i = 1; i < nrow; ++i)
            i_idx[i] = i_idx[0] + i * ncol;  // fill pointer on 1st element of each row
    }
    // copy to integer matrix and sort by row
    for (int i = 0; i < nrow; i++) {
        for (int j = 0; j < ncol; j++) {
            i_idx[i][j] = (int)idx(i,j);
        }
        std::sort(i_idx[i], i_idx[i]+ncol);
    }
    // compute Jaccard
    int *p_i, *p_k, k;
    for (int i = 0; i < nrow; i++) {
        p_i = i_idx[i];
        for (int j = 0; j < ncol; j++) {
            k = i_idx[i][j]-1;
            p_k = i_idx[k];
            // compute intersection of sorted vectors
            int u = 0;
            int i_i = 0, i_k = 0;
            for (;;) {
                if (p_i[i_i] == p_k[i_k]) {
                    u++;
                    i_i++; i_k++;
                    if (i_i >= ncol || i_k >= ncol) break;
                }
                else if (p_i[i_i] < p_k[i_k]) {
                    i_i++;
                    if (i_i >= ncol) break;
                }
                else {
                    i_k++;
                    if (i_k >= ncol) break;
                }
            }
            // compute Jaccard
            if (u < 1) continue;
            weights(r, 0) = i+1;
            weights(r, 1) = k+1;
            weights(r, 2) = u/(4.0*ncol - 2*u);  // symmetrize the graph
            r++;
        }
    }
    // free matrix
    if (nrow) delete [] i_idx[0];
    delete [] i_idx;
    // done
    return weights;
}


// Improve intersection computation by pre-sorting the indices, what requires more memory
// Report a lower number of nearest neighbors, what reduces the graph size and lowers errors on iris
// Author: S. Granjeaud, Date: 24/05/2020

// [[Rcpp::export]]
NumericMatrix jaccard_coeff_4(NumericMatrix idx, IntegerVector ks) {
    int nrow = idx.nrow(), ncol = idx.ncol();
    int nks = ks.length();
    // ks can't be empty
    if (nks == 0)
        stop("ks must be a vector integer with one element at least.");
    int* kx = new int[ncol];  // ncol is the maximal length
    int nkx;
    if (nks == 1 && ks[0] == 0) {
        // if ks is c(0), back to full list of column indices, aka auto mode
        for (int i = 0; i < ncol ; i++) {
            kx[i] = i;
        }
        nkx = ncol;
    } else {
        // subtract 1 from ks indices for C computation, and check boundaries
        bool err_lower = 0, err_greater = 0;
        for (int i = 0; i < nks ; i++) {
            kx[i] = ks[i] - 1;
            if (kx[i] < 0 && !err_lower) err_lower = 1;
            if (kx[i] >= ncol && !err_greater) err_greater = 1;
        }
        nkx = nks;
        if (err_lower || err_greater)
            stop("ks must a vector of integers from 1 to ncol(idx).");
    }
    // Result and its index
    NumericMatrix weights(nrow*nkx, 3);
    int r = 0;
    // allocate memory
    int** i_idx = new int*[nrow];
    i_idx[0] = new int[nrow * ncol];  // mem alloc
    for (int i = 1; i < nrow; ++i)
        i_idx[i] = i_idx[0] + i * ncol;  // fill pointer on 1st element of each row
    // copy to integer matrix and sort by row
    for (int i = 0; i < nrow; i++) {
        for (int j = 0; j < ncol; j++) {
            i_idx[i][j] = (int)idx(i,j);
        }
        std::sort(i_idx[i], i_idx[i]+ncol);
    }
    // compute Jaccard
    int *p_i, *p_k, k;
    for (int i = 0; i < nrow; i++) {
        p_i = i_idx[i];
        for (int j = 0; j < nkx; j++) {
            //k = i_idx[i][ks[j]]-1;
            k = idx(i,kx[j])-1;
            if (i == k) continue;  // avoid loop
            p_k = i_idx[k];
            // compute intersection of sorted vectors
            int u = 0;
            int i_i = 0, i_k = 0;
            for (;;) {
                if (p_i[i_i] == p_k[i_k]) {
                    u++;
                    i_i++; i_k++;
                    if (i_i >= ncol || i_k >= ncol) break;
                }
                else if (p_i[i_i] < p_k[i_k]) {
                    i_i++;
                    if (i_i >= ncol) break;
                }
                else {
                    i_k++;
                    if (i_k >= ncol) break;
                }
            }
            // compute Jaccard
            if (u < 1) continue;  // skip this link if no intersection
            weights(r, 0) = i+1;
            weights(r, 1) = k+1;
            weights(r, 2) = u/(4.0*ncol - 2*u);  // symmetrize the graph
            // symmetrization introduces an extra division by 2, but not clear why this is needed
            r++;
        }
    }
    // free matrix
    delete [] i_idx[0];
    delete [] i_idx;
    delete [] kx;
    // report all nodes (E.K. Becht's issue)
    unsigned long int *reported = new unsigned long int[nrow+1];  // using R indices, 0 unused
    for (int i = 1; i <= nrow; i++)
        reported[i] = 0;
    // loop to check all nodes
    for (int i = 0; i < r; i++) {
        reported[(unsigned long int)weights(i, 0)]++;
        reported[(unsigned long int)weights(i, 1)]++;
    }
    // add all unreported nodes
    for (int i = 1; i <= nrow; i++) {
        if (reported[i] > 0) continue;
        weights(r, 0) = i;
        weights(r, 1) = i;
        weights(r, 2) = 1/2.0;
        // extra division by 2, but not clear why this is needed
        r++;
    }
    delete [] reported;
    // End issue
    // done
    return weights;
}

// DEBUGGING R code, see vignette


R/phenograph.R
#' RphenoGraph clustering
#'
#' R implementation of the PhenoGraph algorithm
#'
#' A simple R implementation of the
#' [PhenoGraph](http://www.cell.com/cell/abstract/S0092-8674(15)00637-6) algorithm, which is a
#' clustering method designed for high-dimensional single-cell data analysis. It works by creating a
#' graph ("network") representing phenotypic similarities between cells by calclating the Jaccard
#' coefficient between nearest-neighbor sets, and then identifying communities using the well known
#' [Louvain method](https://sites.google.com/site/findcommunities/) in this graph.
#'
#' @param data matrix; input data matrix.
#' @param k integer; number of nearest neighbours (default:30).
#' @param directed logical; whether to use a symmetric (FALSE, default) or asymmetric ("directed",
#'   TRUE) graph.
#' @param edge_comb character map multiple edges to single edges; it uses igraph
#'   \code{igraph-attribute-combination}. If graph is directed, there should be no duplicated edge.
#'   If graph is undirected, there are numerous edges that exist in both directions. Those edges
#'   will be combined using the \code{simplify} function, that defines which function to apply to
#'   the weights. Available values are "sum", "prod", "mean", "min", "max"... "first" is used by
#'   default as duplicated links are getting the same weight (same intersection) and is equivalent
#'   to "mean". "sum" and "prod" are reinforcing the weight of links that exist in both directions.
#' @param inc_self logical; whether to include the 1st nearest neighbour, i.e. the data point
#'   itself, when computing the shared neighbours (intersection) with its nearest neighbours. By
#'   default, it is FALSE, as the original Rphenograph implementation. No self link is created,
#'   unless the data point is linked to none of its neighbours because they share no common
#'   neighbours.
#' @param clust_fun function; community detection algorithm. Defaults to cluster_louvain. Louvain is
#'   not implemented for directed graphs. Other options: cluster_walktrap, cluster_spinglass,
#'   cluster_leading_eigen, cluster_edge_betweenness, cluster_fast_greedy, cluster_label_prop.
#' @param verbose logical; verbosity (default=FALSE)
#' @param knn_fun function or character; the function used to search the nearest neighbors. NULL
#'   points to RANN::nn2 as original code, which is configured for exact nearest neighbours (eps=0).
#'   "hnsw" points to RcppHNSW::hnsw_knn as proposed by E. Becht. Any function that returns a matrix
#'   of index to nearest neighbors could be used. NB: the first column contains the point itself and
#'   is removed before building the graph unless inc_self is TRUE.
#' @param knn_report vector of integers; this vector defines the indices within the k nearest
#'   neighbours to report for each data points. This should be a vector of integer from 1 to k. A
#'   single value of 0 indicates that all neighbours (indices from 1 to k) will be reported, as by
#'   default.
#'
#' @return a list contains an igraph graph object for \code{graph_from_data_frame} and a communities
#'   object, for which the most interesting operations are the following:
#'
#'   \item{membership}{returns a numeric vector, one number for each vertex in the graph that was
#'   the input of the community detection.}
#'
#'   \item{length}{returns an integer scalar.}
#'
#'   \item{sizes}{returns a numeric vector.}
#'
#'   \item{modularity}{returns a numeric scalar.}
#'
#'   There are also other operations: \item{algorithm}{returns a character scalar.},
#'   \item{crossing}{returns a logical vector.} \item{is_hierarchical}{returns a logical scalar.},
#'   \item{merges}{returns a two-column numeric matrix.}, \item{cut_at}{returns a numeric vector,
#'   the membership vector of the vertices.}, \item{as.dendrogram}{returns a dendrogram object.},
#'   \item{show_trace}{returns a character vector.}, \item{code_len}{returns a numeric scalar for
#'   communities found with the InfoMAP method and NULL for other methods.}, \item{print}{returns
#'   the communities object itself, invisibly.}, \item{plot}{for communities objects returns NULL,
#'   invisibly.}
#'
#' @details
#'
#' The Python Phenograph algorithm defines a "prune" parameter. prune is a logical that determines
#' whether the graph is simplified (i.e. removing reciprocal edges) by addition (FALSE, default) or
#' multiplication (TRUE). Phenograph is using matrix operations, adjacency matrix of the graph added
#' or multiplied to its transpose. Originally, Rphenograph does not simplify the graph, keeping
#' reciprocal edges. Now Rphenograph offers igraph simplifications, which might lead to different
#' results from Phenograph. In Phenograph, it's not clear what's happening when a link exists in one
#' direction: does product lead to 0 when there is only one link?
#'
#' There is no rational whether to include or not to the data point itself in the list of nearest
#' neighbours. Phenograph does not, but the scran package does. The choice is driven here by
#' \code{inc_self}. Whatever the choice, the graph could be assymmetric when a point A lists B as
#' its k NN, but B does not, because B has k neighbours nearer than A. So A has a link to B, but B
#' has not link to A. When A and B are in the list of each other, there are two reverse links; A and
#' B have the same intersection and thus the same Jaccard coefficient.
#'
#' @references Jacob H. Levine and et.al. Data-Driven Phenotypic Dissection of AML Reveals
#'   Progenitor-like Cells that Correlate with Prognosis. Cell, 2015.
#'
#' @examples
#' library(igraph)
#' iris_unique <- unique(iris) # Remove duplicates
#' data <- as.matrix(iris_unique[,1:4])
#' Rphenograph_out <- Rphenograph(data, k = 45)
#' modularity(Rphenograph_out[[2]])
#' membership(Rphenograph_out[[2]])
#' iris_unique$phenograph_cluster <- factor(membership(Rphenograph_out[[2]]))
#' if(require(ggplot2)) {
#'     ggplot(iris_unique, aes(x=Sepal.Length, y=Sepal.Width, col=Species, shape=phenograph_cluster)) +
#'         geom_point(size = 3) + theme_bw()
#' } else {
#'     with(iris_unique, plot(x=Sepal.Length, y=Sepal.Width,
#'     col=Species, pch=(15:17)[phenograph_cluster],
#'     cex=c(2.5, 2, 1.5)[phenograph_cluster]))
#' }
#'
#' @importFrom igraph graph.data.frame cluster_louvain modularity membership simplify
#' @useDynLib Rphenograph
#'
#' @export
Rphenograph <- function(data, k=30, 
                        directed=FALSE,
                        edge_comb="none",
                        inc_self=FALSE,
                        clust_fun=NULL, 
                        verbose=FALSE, 
                        knn_fun=NULL,
                        knn_report=0)
{
    if(is.data.frame(data))
        data <- as.matrix(data)
    
    if(!is.matrix(data))
        stop("Wrong input data, should be a data frame or matrix!")
    
    if(k<1){
        stop("k must be a positive integer!")
    }else if (k > nrow(data)-2){
        stop("k must be smaller than the total number of points!")
    }

    if(verbose){
        message("\nRun Rphenograph starts:\n", 
                "  -Input data of ", nrow(data)," rows and ", ncol(data), " columns","\n",
                "  -k is set to ", k)
        cat("  Finding nearest neighbors...")
    }
    if(is.null(knn_fun)) knn_fun <- find_neighbors
    if(is.character(knn_fun) && knn_fun == "hnsw")
      if (requireNamespace("RcppHNSW",  quietly = TRUE)) {
        knn_fun <- function(data, k) RcppHNSW::hnsw_knn(data, k=k)$idx
      } else
        stop("RcppHNSW must be installed for using hnsw nearest neighbors search.")
    if(!is.function(knn_fun))
        stop("knn_fun must be a function or a recognized keyword.")
    ## t1 <- system.time(neighborMatrix <- find_neighbors(data, k=k+1)[,-1])
    t1 <- system.time({
      if (inc_self) {
        neighborMatrix <- knn_fun(data, k)
      } else {
        neighborMatrix <- knn_fun(data, k+1)[,-1]
      }
    })

    if(verbose){
      cat("DONE ~",t1[3],"s\n")
      cat("  Compute jaccard coefficient between nearest-neighbor sets...")
    }
    t2 <- system.time({
      links <- jaccard_coeff_4(neighborMatrix, knn_report)
      links <- links[links[,1]>0, ]  # remove unused allocation (unreported zero Jaccard coeff)
      ## Fix if data point goes missing (due to all of its associated jaccard
      ## coefficients being 0 and if it ever appears as another points' nearest
      ## neighbor, the corresponding jaccard coefficient also being 0.
      ## Now done in the C code
    })

    if(verbose){
        cat("DONE ~",t2[3],"s\n", " Build a graph from the weighted links...")
    }

    t3 <- system.time({
      relations <- as.data.frame(links)
      colnames(relations) <- c("from","to","weight")
      g <- graph.data.frame(relations, directed=directed)
      if (directed && edge_comb != "none") {
        cat("  There are typically no duplicated edge when the graph is directed.")
      }
      if (edge_comb == "none") {
        if(verbose)
          cat("\n", " Links are not combined...")
      } else {
        if(verbose)
          cat("\n", " Combining links using", edge_comb, "method...")
        g <- simplify(g, edge.attr.comb = edge_comb, remove.loops = FALSE)
        # Loops are kept to preserve unlinked data points
      }
    })

    if(verbose) {
        cat("DONE ~",t3[3],"s\n")
        cat("  Run ", "clustering on the graph ...")  # TODO: add function name
    }

    # Other community detection algorithms: 
    #    cluster_walktrap, cluster_spinglass, 
    #    cluster_leading_eigen, cluster_edge_betweenness, 
    #    cluster_fast_greedy, cluster_label_prop  
    if(is.null(clust_fun)) clust_fun <- cluster_louvain
    
    t4 <- system.time(community <- clust_fun(g))

    if(verbose){
        cat("DONE ~",t4[3],"s\n")
        
        message("Run Rphenograph DONE, totally takes ", 
                format(sum(c(t1[3],t2[3],t3[3],t4[3])), digits = 5), " s.")
        cat("  Return a community class\n")
        cat("  -Modularity value:", modularity(community), "\n")
        cat("  -Number of clusters:", length(unique(membership(community))), "\n")
    }
    
    return(list(g, community))
}

#' K Nearest Neighbour Search
#'
#' Uses a kd-tree to find the p number of near neighbours for each point in an input/output dataset.
#' 
#' Use the nn2 function from the RANN package, utilizes the Approximate Near Neighbor (ANN) C++ library, 
#' which can give the exact near neighbours or (as the name suggests) approximate near neighbours 
#' to within a specified error bound. For more information on the ANN library please 
#' visit http://www.cs.umd.edu/~mount/ANN/.
#' 
#' @describeIn Rphenograph Uses a kd-tree to find the p number of near neighbours for each point in an input/output dataset.
#' 
#' @return a n-by-k matrix of neighbor indices
#' 
#' @examples
#' iris_unique <- unique(iris) # Remove duplicates
#' data <- as.matrix(iris_unique[,1:4])
#' neighbors <- find_neighbors(data, k=10)
#' 
#' @importFrom RANN nn2
#' @export
find_neighbors <- function(data, k){
    nearest <- nn2(data, data, k, searchtype = "standard")
    return(nearest[[1]])
}
